### ISSUE TITLE
Layout of char

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/unsafe-code-guidelines#101]
Currently we only say that `char` has size `4`. Is there anything we can say about its alignment beyond "implementation-defined" ? 

Also @ubsan [mentioned](https://github.com/rust-lang/unsafe-code-guidelines/issues/9#issuecomment-431048079): 

> I'd argue it'd be useful to be ABI compatible with char32_t) (note: that type only exists in C++).
>
> People have argued that they shouldn't be ABI compatible, since char32_t doesn't have the correctness guarantees Rust's char does; I would argue that it's the same idea as C-like enums in Rust vs enums in C++.

This question is still unresolved.

**Comment by joshtriplett** [comment]
An interesting thought that came up today: there's an alternative representation that might make sense. Since any valid Unicode character can be represented as at most 4 bytes of UTF-8, we *could* represent `char` as a [u8; 4] containing UTF-8.

That would have the interesting property of allowing a function on `char` to return a `&str` without allocating memory (rather than the current `char::encode_utf8`). And `&str` or `String` could return a `char` much more quickly, without having to translate UTF-8 first.

This would have tradeoffs, as well. In exchange for fast character -> UTF-8 and UTF-8 string -> character operations, we'd have to change the implementations of functions that do Unicode character classification and similar, to either convert to a character number first, or provide an optimized mapping from UTF-8 directly to the needed classification (which may be possible in some cases).

I don't know if we'd want to consider this tradeoff, but it seems worth *considering*.

**Comment by Diggsey** [comment]
@joshtriplett that sounds like a useful type, but IMO the extra complexity combined with the extra cost of converting to the corresponding scalar make it not worthwhile.

However, it might be a cool idea as the basis for a "fixed size string" type similar to fixed-size arrays. For example with const generics: `FixedStr<1>` could be equivalent to the type you are describing and have From/To implementations for `char`, whilst also being able to be borrowed as a `&str`.

**Comment by zachs18** [comment]
The docs for `char` now say that "`char` is guaranteed to have the same size, alignment, and function call ABI as `u32` on all platforms." ([link](https://doc.rust-lang.org/nightly/std/primitive.char.html#validity-and-layout), last paragraph of "Validity and Layout" section).

I believe this resolves the questions in the OP about alignment and ABI (with the answer being: "the same as that of `u32`")?

I suppose the question in the OP about compatibility with `char32_t` is still open, but I think that probably is determined by whether `char32_t` is compatible with `u32` on the platform in question.

**Comment by RalfJung** [comment]
Yeah -- and anyway figuring out the mapping to C types is a separate question.
